pkg/binary/envoy/fetch.go

```diff
@@ -15,128 +15,97 @@
 package envoy
 
 import (
-	"archive/tar"
-	"errors"
 	"fmt"
 	"io"
-	"io/ioutil"
 	"net/http"
 	"os"
 	"path/filepath"
 
-	"github.com/mholt/archiver/v3"
 	"github.com/schollz/progressbar/v3"
 
 	"github.com/tetratelabs/getenvoy/pkg/globals"
+	tar "github.com/tetratelabs/getenvoy/pkg/internal"
 	"github.com/tetratelabs/getenvoy/pkg/manifest"
 	"github.com/tetratelabs/getenvoy/pkg/transport"
 )
 
-const envoyLocation = "bin/envoy"
+const binEnvoy = "bin/envoy"
 
 // FetchIfNeeded downloads an Envoy binary corresponding to the given reference and returns a path to it or an error.
 func FetchIfNeeded(o *globals.GlobalOpts, reference string) (string, error) {
-	ref, e := manifest.ParseReference(reference)
-	if e != nil {
-		return "", e
+	ref, err := manifest.ParseReference(reference)
+	if err != nil {
+		return "", err
 	}
 
-	platformDirectory := filepath.Join(o.HomeDir, "builds", ref.Flavor, ref.Version, ref.Platform)
-	envoyPath := filepath.Join(platformDirectory, envoyLocation)
-	stat, e := os.Stat(envoyPath)
-	switch {
-	case os.IsNotExist(e):
-		m, err := manifest.FetchManifest(o.ManifestURL)
-		if err != nil {
-			return "", err
-		}
-		binaryLocation, err := manifest.LocateBuild(ref, m)
-		if err != nil {
-			return "", err
+	platformPath := filepath.Join(o.HomeDir, "builds", ref.Flavor, ref.Version, ref.Platform)
+	envoyPath := filepath.Join(platformPath, binEnvoy)
+	if _, err = os.Stat(envoyPath); os.IsNotExist(err) {
+		if e := os.MkdirAll(platformPath, 0750); e != nil {
+			return "", fmt.Errorf("unable to create directory %q: %w", platformPath, e)
 		}
-		if err = os.MkdirAll(platformDirectory, 0750); err != nil {
-			return "", fmt.Errorf("unable to create directory %q: %w", platformDirectory, err)
+
+		m, e := manifest.FetchManifest(o.ManifestURL)
+		if e != nil {
+			return "", e
 		}
 
-		err = fetchEnvoy(platformDirectory, binaryLocation)
-		if err != nil {
-			return "", err
+		downloadLocationURL, e := manifest.LocateBuild(ref, m)
+		if e != nil {
+			return "", e
 		}
-	case e != nil:
-		return "", fmt.Errorf("invalid Envoy binary at %q: %w", envoyPath, e)
-	default:
-		fmt.Printf("%v is already downloaded\n", ref)
-		if stat.Mode()&0111 == 0 {
-			return "", fmt.Errorf("envoy binary not executable: %s", envoyPath)
+
+		fmt.Fprintln(o.Out, "downloading", downloadLocationURL) //nolint
+		if e := untarEnvoy(platformPath, downloadLocationURL, o.Out); e != nil {
+			return "", e
 		}
+	} else {
```

Comment:
I think you missed a case of `err != nil && !os.IsNotExist(err)`

Reply:
yep this was a miss. we need a boiler-plate way to make os.Stat err.. I fixed the flow logic and added a TODO for now